project without tests

Garret-vin opened the pull request:
omdb api
spring batch
swagger

Garret-vin committed:
project without tests
omdb api
spring batch
swagger

Garret-vin committed:
change method name in MovieService
isNotPresent -> existsInDb

Garret-vin committed:
add Language entity
add get from db by language controller

Garret-vin committed:
add api to get by votes
add api to get by rating
small fix in entities

Garret-vin committed:
add JUnit tests to MovieUtil class

Garret-vin committed:
add omdbClient mock tests
add part of mock tests MovieService

Garret-vin committed:
change travis settings

Garret-vin committed:
change travis settings

Garret-vin committed:
add custom exceptions
add tests to movieServiceImpl
refactoring classes + pom

Garret-vin committed:
changed ignore settings for codecov

mmarusich commented on codecov.yml:
```diff
@@ -1,6 +1,16 @@
-    - "com.garret.movies.moviestore.MovieStoreApplication.java"
\ No newline at end of file
+    - "dao/.*"
+    - "movie-service/src/main/java/com/garret/movies/service/api/.*"
```
@Garret-vin @Dominys Instead of exclusion here configure exclusion for jacoco coverage report

mmarusich commented on dao/pom.xml:
```diff
@@ -0,0 +1,36 @@
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <optional>true</optional>
```
scope =  compile only, we need lombok only for compilation

mmarusich commented on dao/src/main/java/com/garret/movies/dao/entity/Actor.java:
```diff
@@ -0,0 +1,18 @@
+
+@Data
+@Entity
+@NoArgsConstructor
```
@Garret-vin @Data annotation already includes RequiredArgsConstructor and in your case this is empty constructor, so  you can remove NoArgsConstructor

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    @Transactional
+    public Movie save(@NonNull Movie movie) {
+        Optional<Movie> optionalMovie = movieRepository.findByImdbId(movie.getImdbId());
+        if (optionalMovie.isPresent()) {
```
@Garret-vin In most of cases Optional::get can be avoided, try to use build in mehods of optional to transform data.
In our case whole this method code can be replaced with:

`
return movieRepository.findByImdbId(movie.getImdbId()).orElseGet(() -> movieRepository.save(movie));
`

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    public List<Movie> saveAll(@NonNull List<Movie> movies) {
+        movieRepository.saveAll(movies);
+        log.info(movies.size() + " movies saved to database");
+        return movies;
```
@Garret-vin Its better to return result of save method as result movies will contain ids which was generated during persist. Returning input data not super useful as you already have it in place where you call this service

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    @Transactional(readOnly = true)
+    public List<Movie> getAll() {
+        return ImmutableList.copyOf(movieRepository.findAll());
```
List.of

Garret-vin replied:
@mmarusich I can't use List.of() because Iterable is inconvertible type to List

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<Movie> getAllByYear(@NonNull Integer year) {
```
@Garret-vin If its NotNull so why you passing Object? Why not pass primitive int?

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    public List<Movie> getAllByYear(@NonNull Integer year) {
+        String start = year + "-01-01";
+        String end = year + "-12-31";
+        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd");
```
@Garret-vin If you go inside SimpleDateFormat you will see that during initialization it creates regex and initialize it which takes time and resources. As you have static format template its better to move it in final class variable and init only once instead of doing that each time

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+            Date endDate = formatter.parse(end);
+            return movieRepository.findAllByReleasedBetween(startDate, endDate);
+        } catch (ParseException e) {
+            log.error("Incorrect date in MovieServiceImpl.class throws from method: getAllByYear()", e);
```
@Garret-vin Try providing info to logs/error which will help you quickly fins issue, in this example it will be useful to print year value

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+        boolean isDeleted = false;
+        if (movieRepository.existsById(id)) {
+            movieRepository.deleteById(id);
+            isDeleted = true;
```
@Garret-vin Do you rely need this? You doing additional call to DB just to check if it was deleted, why you care?

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientImpl.java:
```diff
@@ -0,0 +1,75 @@
+@AllArgsConstructor
+public class OmdbClientImpl implements OmdbClient {
+
+    private OmdbApi omdbApi;
```
@Garret-vin Add final and use @RequiredArgsConstuctor. Otherwise if you will add some class variable for internal use it will be added to constructor as well which you don't need

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientImpl.java:
```diff
@@ -0,0 +1,75 @@
+    public List<Movie> searchMovies(@NonNull String title) {
+        try {
+            List<OmdbVideoBasic> videoList = ListUtils.emptyIfNull(omdbApi.search(title).getResults());
+            return ListUtils.emptyIfNull(videoList.stream()
```
@Garret-vin emptyIfNull will never happen as stream collect can't return null

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientImpl.java:
```diff
@@ -0,0 +1,75 @@
+        } catch (OMDBException e) {
+            log.error("Invalid parameters", e);
+        }
+        return Collections.emptyList();
```
@Garret-vin This is expected behaviour? You will not be able to react on errors in service which will call search. If you providing API you also should provide information about errors

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/config/OmdbConfig.java:
```diff
@@ -0,0 +1,18 @@
+    private String API_KEY;
+
+    @Bean
+    public OmdbApi omdbApi() {
```
@Garret-vin You can pass API key as argument to this method:

` public OmdbApi omdbApi(@Value("${api_key}") String key) `

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/util/MovieUtil.java:
```diff
@@ -0,0 +1,89 @@
+        }
+        try {
+            String pattern = "dd MMM yyyy";
+            SimpleDateFormat formatter = new SimpleDateFormat(pattern, Locale.ENGLISH);
```
@Garret-vin See previous comment regarding date formats

mmarusich commented on web/src/main/java/com/garret/movies/MovieApplication.java:
```diff
@@ -0,0 +1,15 @@
+
+@SpringBootApplication
+@EnableBatchProcessing
+@PropertySource(value = {"application.properties", "omdb.properties", "dao.properties"})
```
@Garret-vin Application properties by default already in context, about others please add property source on top of beans which holds this config

mmarusich commented on web/src/main/java/com/garret/movies/config/SpringBatchConfig.java:
```diff
@@ -0,0 +1,68 @@
+public class SpringBatchConfig {
+
+    private MovieService movieService;
+    private OmdbClient omdbClient;
```
again final and required

mmarusich commented on web/src/main/java/com/garret/movies/config/SpringBatchConfig.java:
```diff
@@ -0,0 +1,68 @@
+    @Bean
+    @StepScope
+    public ItemReader<Movie> itemReader(@Value("#{jobParameters[title]}") String title) {
+        final List<Movie> newMovies = omdbClient.searchMovies(title)
```
@Garret-vin Not sure I got how this suppose to work, you reading data when bean will be initialized

mmarusich commented on web/src/main/java/com/garret/movies/controller/MovieDatabaseController.java:
```diff
@@ -0,0 +1,61 @@
+
+@RestController
+@AllArgsConstructor
+@RequestMapping("/db")
```
@Garret-vin Please read about Rest API best practices and naming conversions https://restfulapi.net/resource-naming/, https://blog.mwaysolutions.com/2014/06/05/10-best-practices-for-better-restful-api/

mmarusich commented on web/src/main/java/com/garret/movies/controller/MovieOmdbController.java:
```diff
@@ -0,0 +1,38 @@
+        JobExecution jobExecution = jobLauncher.run(job, jobParameters);
+        log.info("Job execution status: " + jobExecution.getStatus());
+
+        return jobExecution.getStatus();
```
@Garret-vin This implementations means synchronize job execution from HTTP call, so you will block call until your job will finish, its not right as you probably will have HTTP timeouts and in addition blocking thread always bad

Garret-vin committed:
was changed:
ignore settings for codecov
settings for jacoco plugin
MovieDatabaseController optimized
add async to spring batch
other small fixes

mmarusich commented on movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java:
```diff
@@ -22,28 +22,25 @@
+        return movieRepository.findByImdbId(movie.getImdbId()).orElseGet(() -> {
+            log.info("Saving movie " + movie.getTitle());
+            return movieRepository.save(movie);
+        });
```
:+1:

Garret-vin committed:
add jooq dependency+config plugin
add @Async to omdbController
